lib/SiteClassifier.jsm

```diff
@@ -0,0 +1,62 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {RemoteSettings} = ChromeUtils.import("resource://services-settings/remote-settings.js", {});
+
+// Returns whether the passed in params match the criteria.
+// To match, they must contain all the params specified in criteria and the values
+// must match if a value is provided in criteria.
+function _hasParams(criteria, params) {
+  for (let param of criteria) {
+    const val = params.get(param.key);
+    if (val === null || (param.value && param.value !== val)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * classifySite
+ * Classifies a given URL into a category based on classification data from RemoteSettings.
+ *
+ */
+async function classifySite(url, RS = RemoteSettings) {
+  let category = "other";
+  let parsedURL;
+
+  // Try to parse the url.
+  for (let _url of [url, `http://${url}`]) {
+    try {
+      parsedURL = new URL(_url);
+      break;
+    } catch (e) {}
+  }
+
+  if (parsedURL) {
+    // If we parsed successfully, find a match.
+    const hostname = parsedURL.hostname.replace(/^www\./i, "");
+    const params = parsedURL.searchParams;
+    const siteTypes = await RS("sites-classification").get();
+    const sortedSiteTypes = siteTypes.sort((x, y) => y.weight - x.weight);
```

Comment:
If we check the type of URL in the ascending order in weight, and return immediately once we find a match. Will we miss the chance to get the more specific matches with higher weight? For example, say we have two types defined as:

```js
{
  "type": "host-match",
  "critiria": {
     "hostname": "foobar.com"
   }
  "weight": 100,
}

{
  "type": "host-and-params-must-match",
  "critiria": {
     "hostname": "foobar.com",
     "params": ...
   }
  "weight": 200,
}
```

Given the `foobar.com/must/match`, it'll return type as `host-match`. Shall we check the type in descending order then?